### PR TITLE
Remove `BaseBreadcrumbMenuBlockService::getRootMenu`

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this is a major upgrade path.

## Changelog

```markdown
### Changed
- `BaseBreadcrumbMenuBlockService::getMenu()` method contains the root menu

### Removed
- `BaseBreadcrumbMenuBlockService::getRootMenu()` method
```

### CODE REVIEW

File: src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php [hunk 1]
```diff
@@ -75,33 +75,18 @@ public function configureSettings(OptionsResolver $resolver): void
             'menu_class' => 'list-group',
             'children_class' => 'list-group-item',
             'menu_template' => '@SonataSeo/Block/breadcrumb.html.twig',
-            'include_homepage_link' => true,
+            'include_homepage_link' => \true,
```
Comment (VincentLanglet) [inline src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php]: this is weird

File: src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php [hunk 2]
```diff
@@ -75,33 +75,18 @@ public function configureSettings(OptionsResolver $resolver): void
...
     {
         $settings = $blockContext->getSettings();
-        /*
-         * @todo : Use the router to get the homepage URI
-         */
 
         $menu = $this->factory->createItem('breadcrumb');
         $menu->setChildrenAttribute('class', 'breadcrumb');
-        $menu->setCurrent(true);
+        $menu->setCurrent(\true);
```
Comment (VincentLanglet) [inline src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php]: same

File: tests/Block/Breadcrumb/BreadcrumbTest.php [hunk 3]
```diff
@@ -93,7 +87,7 @@ public function testDefaultSettings(): void
             'menu_class' => 'list-group',
             'children_class' => 'list-group-item',
             'menu_template' => '@SonataSeo/Block/breadcrumb.html.twig',
-            'include_homepage_link' => true,
+            'include_homepage_link' => \true,
```
Comment (VincentLanglet) [inline tests/Block/Breadcrumb/BreadcrumbTest.php]: true

File: src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php [hunk 4]
```diff
@@ -111,6 +96,13 @@ final protected function getRootMenu(BlockContextInterface $blockContext): ItemI
         return $menu;
     }
 
+    final protected function getFactory(): FactoryInterface
+    {
+        return $this->factory;
+    }
+
+    abstract protected function getContext(): string;
```
Comment (VincentLanglet) [inline src/Block/Breadcrumb/BaseBreadcrumbMenuBlockService.php:104]: You may add the changes of https://github.com/sonata-project/SonataSeoBundle/pull/622/files in this PR ?

File: tests/Block/Breadcrumb/BreadcrumbTest.php (hunk 3 above)
Comment (core23) [reply tests/Block/Breadcrumb/BreadcrumbTest.php]: WTF?

Looks like my local cs-fixer runs in strict mode oO

**Review by VincentLanglet — DISMISSED** [review]

File: UPGRADE-3.0.md [hunk 5]
```diff
@@ -1,6 +1,30 @@
 UPGRADE FROM 2.X to 3.0
 =======================
 
+## Refactored `BaseBreadcrumbMenuBlockService`
+
+The class was slightly refactored, so some extension points became final.
+
+The `getRootMenu` method was removed and the logic was moved to the `getMenu` method. You need to call the parent method if you rely on the old logic:
```
Comment (phansys) [inline UPGRADE-3.0.md]:
```suggestion
The `getRootMenu()` method was removed and the logic was moved to the `getMenu()` method. You need to call the parent method if you rely on the old logic:
```

File: UPGRADE-3.0.md [hunk 6]
```diff
@@ -1,6 +1,30 @@
...
+    // come custom logic
+}
+```
+After:
+```php
+protected function getMenu(BlockContextInterface $blockContext): {
+    parent::getMenu($blockContext);
+
+    // come custom logic
+
```
Comment (phansys) [inline UPGRADE-3.0.md]:
```suggestion
```

File: UPGRADE-3.0.md [hunk 7]
```diff
@@ -1,6 +1,30 @@
...
+Before:
+```php
+protected function getMenu(BlockContextInterface $blockContext) {
+    // come custom logic
+}
+```
+After:
+```php
+protected function getMenu(BlockContextInterface $blockContext): {
```
Comment (phansys) [inline UPGRADE-3.0.md]:
```suggestion
protected function getMenu(BlockContextInterface $blockContext): [RETURN_TYPE] {
```

**Review by phansys — CHANGES_REQUESTED** [review]

**Review by VincentLanglet — DISMISSED** [review]

File: UPGRADE-3.0.md [hunk 8]
```diff
@@ -1,6 +1,28 @@
...
+After:
+```php
+protected function getMenu(BlockContextInterface $blockContext): ItemInterface {
+    parent::getMenu($blockContext);
+    // come custom logic
+}
+```
+
+
+
```
Comment (phansys) [inline UPGRADE-3.0.md]:
```suggestion
```

**Review by phansys — CHANGES_REQUESTED** [review]

**Review by phansys — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
thanks

**Pull request merged**